File: everything_parsing/09-input-buffering/js/buffer.mjs
// Module 09 — the two-buffer input scheme from the Dragon Book (§3.2).
//
//   - two halves of `halfSize` characters, filled from the source on demand
//   - a SENTINEL slot after each half: the inner loop tests ONLY `c === SENTINEL`,
//     one compare that means "reload the other half" or (as EOF) "stop"
//   - lexemeBegin marks the lexeme start; forward scans ahead
//   - retract is forward--  : O(1)
//
// Cost: each source char is copied into a half exactly once → ~N/halfSize loads.
// Limit: a lexeme must fit in one half (so retract never crosses the boundary).

export const EOF = -1;
const SENTINEL = -2;

export class Buffer {
  constructor(source, halfSize) {
    this.source = source;
    this.halfSize = halfSize;
    this.data = new Int32Array(2 * (halfSize + 1));
    this.sourcePos = 0;
    this.consumedCount = 0;
    this.bufferLoads = 0;
    this.forwardReads = 0;
    this.fillHalf(0);
    this.forward = 0;
    this.lexemeBegin = 0;
  }

  halfStart(half)     { return half * (this.halfSize + 1); }
  sentinelIndex(half) { return this.halfStart(half) + this.halfSize; }
  whichHalf(idx)      { return idx < this.halfSize + 1 ? 0 : 1; }

  fillHalf(half) {
    this.bufferLoads++;
    const base = this.halfStart(half);
    const n = Math.min(this.halfSize, this.source.length - this.sourcePos);
    for (let i = 0; i < n; i++) this.data[base + i] = this.source.charCodeAt(this.sourcePos + i);
    this.sourcePos += n;
    if (n < this.halfSize) this.data[base + n] = EOF;
    else this.data[this.sentinelIndex(half)] = SENTINEL;
  }

  crossBoundary() {
    const other = 1 - this.whichHalf(this.forward);
    this.fillHalf(other);
    this.forward = this.halfStart(other);
  }

  peek() {
    for (;;) {
      const c = this.data[this.forward];
      if (c === SENTINEL) { this.crossBoundary(); continue; }
      return c;
    }
  }

  advance() {
    for (;;) {
      const c = this.data[this.forward];
      this.forwardReads++;
      if (c === SENTINEL) { this.crossBoundary(); continue; }
      if (c === EOF) return EOF;
      this.forward++;
      this.consumedCount++;
      return c;
    }
  }

  mark() { this.lexemeBegin = this.forward; }

  retract(n) {
    for (let i = 0; i < n; i++) {
      this.forward--;
      this.consumedCount--;
      if (this.forward < this.halfStart(this.whichHalf(this.forward)) || this.data[this.forward] === SENTINEL)
        throw new Error("retract crossed the halfway boundary (lexeme too long)");
    }
  }

  lexeme() {
    let s = "";
    let i = this.lexemeBegin;
    while (i !== this.forward) {
      const c = this.data[i];
      if (c === SENTINEL) { i = this.halfStart(1 - this.whichHalf(i)); continue; }
      if (c === EOF) break;
      s += String.fromCharCode(c);
      i++;
    }
    return s;
  }

  atEof()      { return this.peek() === EOF; }
  consumed()   { return this.consumedCount; }
  currentHalf() { return this.whichHalf(this.forward); }
}

File: everything_parsing/09-input-buffering/java/Buffer.java
/**
 * Module 09 — the two-buffer input scheme from the Dragon Book (§3.2).
 *
 * The problem: a scanner reads a character, decides it needs to look further,
 * reads more, then sometimes RETRACTS (a lexeme actually ended one character
 * back). The real constraint is that input arrives in blocks and you can't hold
 * the whole file.
 *
 * The scheme:
 *   - two halves of `halfSize` characters each, filled from the source on demand
 *   - a SENTINEL slot after each half. The sentinel is a value that cannot occur
 *     in the input, so the inner loop tests ONLY `c == SENTINEL` — one compare
 *     that means either "reload the other half" or (as EOF) "stop", instead of
 *     testing `forward == end` on every character.
 *   - `lexemeBegin` marks the start of the current lexeme; `forward` scans ahead.
 *   - retract is `forward--`: O(1).
 *
 * Cost: each source character is copied into a half exactly once, so N chars
 * cost ~N/halfSize buffer loads — O(1) amortized per character.
 *
 * Known limit (also the Dragon Book's): a lexeme must fit within one half, so
 * you never need to retract across the halfway boundary into an overwritten half.
 */
public final class Buffer {

    public static final int EOF       = -1;   // true end of input
    private static final int SENTINEL = -2;   // "this slot is a half boundary; reload"

    private final String source;
    private final int halfSize;
    private final int[] data;      // [ half0 (halfSize) | S | half1 (halfSize) | S ]
    private int sourcePos = 0;     // next source char not yet copied into a half
    private int forward;           // index into data
    private int lexemeBegin;
    private int consumed = 0;      // real characters advanced past

    // instrumentation for the tutorial
    int bufferLoads = 0;
    int forwardReads = 0;         // one per `data[forward]` inspection in advance()

    public Buffer(String source, int halfSize) {
        this.source = source;
        this.halfSize = halfSize;
        this.data = new int[2 * (halfSize + 1)];
        fillHalf(0);
        forward = 0;
        lexemeBegin = 0;
    }

    private int halfStart(int half)     { return half * (halfSize + 1); }
    private int sentinelIndex(int half) { return halfStart(half) + halfSize; }
    private int whichHalf(int idx)      { return idx < (halfSize + 1) ? 0 : 1; }

    private void fillHalf(int half) {
        bufferLoads++;
        int base = halfStart(half);
        int n = Math.min(halfSize, source.length() - sourcePos);
        for (int i = 0; i < n; i++) data[base + i] = source.charAt(sourcePos + i);
        sourcePos += n;
        if (n < halfSize) data[base + n] = EOF;             // source ran out inside this half
        else               data[sentinelIndex(half)] = SENTINEL; // half full: boundary
    }

    private void crossBoundary() {
        int other = 1 - whichHalf(forward);
        fillHalf(other);
        forward = halfStart(other);
    }

    /** The character at `forward` without moving. Char or EOF, never SENTINEL. */
    public int peek() {
        while (true) {
            int c = data[forward];
            if (c == SENTINEL) { crossBoundary(); continue; }
            return c;
        }
    }

    /** Return the character at `forward`, then move forward. At EOF, returns EOF and stays. */
    public int advance() {
        while (true) {
            int c = data[forward];
            forwardReads++;
            if (c == SENTINEL) { crossBoundary(); continue; }
            if (c == EOF) return EOF;
            forward++;
            consumed++;
            return c;
        }
    }

    /** Begin a new lexeme at the current position. */
    public void mark() { lexemeBegin = forward; }

    /** Push `forward` back by n characters. n must keep us within the current half. */
    public void retract(int n) {
        for (int i = 0; i < n; i++) {
            forward--;
            consumed--;
            if (forward < halfStart(whichHalf(forward)) || data[forward] == SENTINEL)
                throw new IllegalStateException("retract crossed the halfway boundary (lexeme too long)");
        }
    }

    /** The text from lexemeBegin to forward, stitched across the halfway split. */
    public String lexeme() {
        StringBuilder sb = new StringBuilder();
        int i = lexemeBegin;
        while (i != forward) {
            int c = data[i];
            if (c == SENTINEL) { i = halfStart(1 - whichHalf(i)); continue; }
            if (c == EOF) break;
            sb.append((char) c);
            i++;
        }
        return sb.toString();
    }

    public boolean atEof()   { return peek() == EOF; }
    public int consumed()    { return consumed; }
    public int currentHalf() { return whichHalf(forward); }
}
